### ISSUE TITLE
Feature request: Simple external data source

### ISSUE THREAD

**Issue by canadiannomad** [hashicorp/terraform#17339]
Most of the time I'm writing an external data source, all I want is the string output of a command.  No JSON parsing, no fancy objects, just the value.
On more than one occasion I find myself wanting to run a simple command that I could use as a trigger, that I could use for a hash value, that I could use for a conditional, output a generated password, etc.
In all these cases, I find it cumbersome and error-prone to try to wrap it in JSON before handing it to terraform(eg if there might be quotes in it).
Perhaps we could add a flag that would allow simple strings to be received from any command?
Perhaps when such a flag is set the data source could merely return an object with {stdout, stderr, retval} and leave it to us to interpret it?

**Comment by apparentlymart** [comment]
Hi @canadiannomad!

We've discussed before the idea of a `local_exec` data source that would behave like you describe here. 

The `external` provider was an experiment in allowing more tightly-integrated external programs to be written, particular given that it's hard to parse complex command output in Terraform config, but indeed it does make it hard to use when you're running a program that just produces a simple string value that you want to use verbatim.

The hypothetical `local_exec` would be explicitly a low-level data source that leaves the caller to figure out what to do with the result:

```
# Not yet implemented
data "local_exec" "example" {
  command     = ["/bin/example"]
  working_dir = "/tmp/example"
}

output "example" {
  value = "${data.local_exec.example.stdout}"
}
```

By default I would expect this to still fail if the program exits with a non-successful status, since there's no control flow capabilities in Terraform config and so having to conditionally deal with failure everywhere else in the configuration would be unmanageable in most cases. We could consider an opt-in setting to still "succeed" when the command exits unsuccessfully, but I'd prefer to see a good example of how that would be used (including a config example that shows how it would be expressed in config).

The Terraform team at HashiCorp won't be able to work on this in the near future due to our focus being elsewhere, but we'd be happy to review a pull request if you or someone else has the time and motivation to implement it. I'm going to have our bot move this issue into the repository for the "local" provider since that is where this work would take place.
